p2p/transport/webrtc/udpmux/muxed_connection.go

```diff
@@ -0,0 +1,205 @@
+package udpmux
+
+import (
+	"context"
+	"errors"
+	"net"
+	"sync"
+	"time"
+
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+var _ net.PacketConn = &muxedConnection{}
+
+var (
+	errAlreadyClosed = errors.New("already closed")
+)
+
+const maxPacketsInQueue = 128
+
+// muxedConnection provides a net.PacketConn abstraction
+// over packetQueue and adds the ability to store addresses
+// from which this connection (indexed by ufrag) received
+// data.
+type muxedConnection struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	pq      *packetQueue
+	ufrag   string
+	address *string
+	mux     *udpMux
+}
+
+func newMuxedConnection(mux *udpMux, ufrag string) *muxedConnection {
+	ctx, cancel := context.WithCancel(mux.ctx)
+	return &muxedConnection{
+		ctx:    ctx,
+		cancel: cancel,
+		pq:     newPacketQueue(),
+		ufrag:  ufrag,
+		mux:    mux,
+	}
+}
+
+func (conn *muxedConnection) GetAddress() (string, bool) {
+	if conn.address == nil {
+		return "", false
+	}
+	return *conn.address, true
+}
+
+func (conn *muxedConnection) SetAddress(s string) {
+	if conn.address != nil && *conn.address != s {
+		log.Debugf("address changed for ufrag %s: %s -> %s", conn.ufrag, *conn.address, s)
+	}
+	conn.address = &s
+}
+
+func (conn *muxedConnection) Push(buf []byte, addr net.Addr) error {
+	return conn.pq.Push(buf, addr)
+}
+
+// Close implements net.PacketConn
+func (conn *muxedConnection) Close() error {
+	_ = conn.closeConnection()
+	conn.mux.RemoveConnByUfrag(conn.ufrag)
+	return nil
+}
+
+// LocalAddr implements net.PacketConn
+func (conn *muxedConnection) LocalAddr() net.Addr {
+	return conn.mux.socket.LocalAddr()
+}
+
+// ReadFrom implements net.PacketConn
+func (conn *muxedConnection) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+	return conn.pq.Pop(conn.ctx, p)
+}
+
+// SetDeadline implements net.PacketConn
+func (*muxedConnection) SetDeadline(t time.Time) error {
+	// no deadline is desired here
+	return nil
+}
+
+// SetReadDeadline implements net.PacketConn
+func (*muxedConnection) SetReadDeadline(t time.Time) error {
+	// no read deadline is desired here
+	return nil
+}
+
+// SetWriteDeadline implements net.PacketConn
+func (*muxedConnection) SetWriteDeadline(t time.Time) error {
+	// no write deadline is desired here
+	return nil
+}
```

Comment:
Hehe.. Go.... Sure :) In this case it's already going to not compile either way if you do not have this though, but sure, I can also add an explicit assertion :) I've been in the Go community since around 12 years ago, at some point people started liking these "assertions" :P

Reply:
Also I usually type them like `var _ net.PacketConn = (*muxedConnection)(nil)` if I need to use them. I usually only implement these for public APIs where I want to make sure I do not break people even though I do not advertise and use the interface myself. But sure, I added it!